pkg/endpoint/policy.go

```diff
@@ -923,6 +931,74 @@ func (e *Endpoint) runIdentityToK8sPodSync() {
 	)
 }
 
+// FormatGlobalEndpointID returns the global ID of endpoint in the format
+// / <global ID Prefix>:<cluster name>:<node name>:<endpoint ID> as a string.
+func (e *Endpoint) FormatGlobalEndpointID() string {
+	nodeIdentity, _ := node.GetLocalNode()
+	metadata := []string{CiliumGlobalIdPrefix, ipcache.AddressSpace, nodeIdentity.Name, strconv.Itoa(int(e.ID))}
+	return strings.Join(metadata, ":")
+}
+
+// This synchronizes the key-value store with a mapping of the endpoint's IP
+// with the numerical ID representing its security identity.
+func (e *Endpoint) runIPIdentitySync(endpointIP addressing.CiliumIP) {
+
+	if endpointIP == nil {
+		return
+	}
+
+	addressFamily := endpointIP.GetFamilyString()
+	ipKey := path.Join(ipcache.IPIdentitiesPath, ipcache.AddressSpace, endpointIP.String())
+
+	e.controllers.UpdateController(fmt.Sprintf("sync-%s-identity-mapping (%d)", addressFamily, e.ID),
+		controller.ControllerParams{
+			DoFunc: func() error {
+
+				e.Mutex.RLock()
+
+				if e.state == StateDisconnected || e.state == StateDisconnecting {
+					log.WithFields(logrus.Fields{logfields.EndpointState: e.state}).
+						Debugf("not synchronizing endpoint IP with kvstore due to endpoint state")
+					e.Mutex.RUnlock()
+					return nil
+				}
+
+				if e.SecurityIdentity == nil {
+					e.Mutex.RUnlock()
+					return nil
+				}
+				//identityValue := e.SecurityIdentity.ID.StringID()
+				ipIDPair := identityPkg.IPIdentityPair{
+					IP:       endpointIP.IP(),
+					ID:       e.SecurityIdentity.ID,
+					Metadata: e.FormatGlobalEndpointID(),
+				}
+
+				// Release lock as we do not want to have long-lasting key-value
+				// store operations resulting in lock being held for a long time.
+				e.Mutex.RUnlock()
+
+				marshaledIPIDPair, err := json.Marshal(ipIDPair)
+				if err != nil {
+					return err
+				}
+
+				if err := kvstore.Update(ipKey, marshaledIPIDPair, true); err != nil {
+					return fmt.Errorf("unable to add endpoint IP '%s' to identity '%s': %s", ipKey, marshaledIPIDPair, err)
+				}
+				return nil
+			},
+			StopFunc: func() error {
+				if err := kvstore.Delete(ipKey); err != nil {
```

Comment:
Not related to your PR directly but `StopFunc()` is not being retried. Your code is an example of where it would make sense to retry for at least a bit. I'm still a bit undecided as the lease will cause it to be removed if the agent reconnects and creates a new lease.

Reply:
I thought the lease was created on a per-key basis? I.e., if the entry fails to be deleted, and there is no such endpoint once the agent reconnects, after the lease expires, the key-value pair will get deleted from the key-value store. Is this assumption incorrect?

Reply:
Lease is created per client and attached to all keys which request it. Lease renewal thus applies to all keys of an agent right now.